qadence/backends/pyqtorch/convert_ops.py

```diff
@@ -313,6 +316,19 @@ def __init__(
             self.register_buffer("hmat", hmat)
             self._hamiltonian = lambda self, values: self.hmat
 
+        if (
+            isinstance(block.generator, AbstractBlock)
+            and not block.is_parametric
+            and block.generator._is_diag_pauli
+        ):
+            hmat = block_to_diagonal(
+                block.generator, qubit_support=self.qubit_support, use_full_support=False
+            )
+
+            hmat = hmat.permute(1, 2, 0)
+            self.register_buffer("hmat", hmat)
+            self._hamiltonian = lambda self, values: self.hmat
```

Comment:
Note that here `self.hmat` should now be a single vector instead of a matrix, so I suspect you still need to change the code when this function is called to account for that. You should look at the `unitary` and `_unitary` methods, and the `_evolve_diag_operator` function.

As for the test, like @dominikandreasseitz, first let's see if all current tests pass. It might be enough for this change, this it is just a refactoring of something that was hopefully already tested. But I will take a look to double-check.

Reply:
@jpmoutinho Thank you, I will make the corrections